File: Stack/Min Stack/MinStack.cpp
class MinStack
{
private:
    struct StackElement
    {
        int value;
        int minValue;

        StackElement(int val, int minVal) : value(val), minValue(minVal) {}
    };
    
    stack<StackElement> minStack;

public:
    MinStack() {}

    void push(int val)
    {
        if (minStack.empty())
        {
            minStack.push(StackElement(val, val));
        }
        else
        {
            int newMin = std::min(val, minStack.top().minValue);
            minStack.push(StackElement(val, newMin));
        }
    }

    void pop()
    {
        if (!minStack.empty())
        {
            minStack.pop();
        }
    }

    int top()
    {
        return minStack.top().value;
    }

    int getMin()
    {
        return minStack.top().minValue;
    }
};

File: Stack/Min Stack/MinStack.java
import java.util.Stack;

class MinStack {
    Stack<PairStack> stack;

    public MinStack() {
        stack = new Stack<>();
    }

    public void push(int val) {
        if (stack.isEmpty()) {
            stack.push(new PairStack(val, val));
        } else {
            int currentMinInStack = stack.peek().min;
            stack.push(new PairStack(val, Math.min(currentMinInStack, val)));
        }
    }

    public void pop() {
        if (!stack.isEmpty()) {
            stack.pop();
        }
    }

    public int top() {
        return stack.peek().val;
    }

    public int getMin() {
        return stack.peek().min;
    }

    class PairStack {
        int val;
        int min;

        PairStack(int val, int min) {
            this.val = val;
            this.min = min;
        }
    }
}

